src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -213,37 +213,35 @@ private ImmutableHashSet<ProjectId> GetProjectsThatDirectlyDependOnThisProject_N
 
     private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeReverseReferencesMap()
     {
-        using var _ = PooledDictionary<ProjectId, ImmutableHashSet<ProjectId>.Builder>.GetInstance(out var reverseReferencesMap);
-
+        using var _1 = PooledDictionary<ProjectId, ImmutableHashSet<ProjectId>.Builder>.GetInstance(out var reverseReferencesMapBuilders);
         foreach (var (projectId, references) in _referencesMap)
         {
             foreach (var referencedId in references)
             {
-                if (!reverseReferencesMap.TryGetValue(referencedId, out var builder))
+                if (!reverseReferencesMapBuilders.TryGetValue(referencedId, out var builder))
                 {
                     builder = s_reverseReferencesBuilderPool.Allocate();
-                    reverseReferencesMap.Add(referencedId, builder);
+                    reverseReferencesMapBuilders.Add(referencedId, builder);
                 }
 
                 builder.Add(projectId);
             }
         }
 
-        return reverseReferencesMap
-            .Select(kvp =>
-            {
-                var builder = kvp.Value;
+        using var _2 = PooledDictionary<ProjectId, ImmutableHashSet<ProjectId>>.GetInstance(out var reverseReferencesMap);
+        foreach (var (projectId, builder) in reverseReferencesMapBuilders)
+        {
+            // Realize an ImmutableHashSet from the builder
+            var reverseReferencesForProject = builder.ToImmutableHashSet();
 
-                // Realize an ImmutableHashSet from the builder
-                var reverseReferencesForProject = builder.ToImmutableHashSet();
+            // Clear out the builder and release it back to the pool
+            builder.Clear();
+            s_reverseReferencesBuilderPool.Free(builder);
 
-                // Clear out the builder and release it back to the pool
-                builder.Clear();
-                s_reverseReferencesBuilderPool.Free(builder);
+            reverseReferencesMap.Add(projectId, reverseReferencesForProject);
+        }
 
-                return KeyValuePairUtil.Create(kvp.Key, reverseReferencesForProject);
-            })
-            .ToImmutableDictionary();
+        return reverseReferencesMap.ToImmutableDictionary();
```

Comment:
poooled dict, to immutable dict?  shouldn't this just be an ImmutableDictBuilder?

Reply:
It could be, but the builder is a class itself, so this avoids that allocation. I can switch to that though if you prefer.

Reply:
teh main concern here is that to go from Dict to ImmutableDict you are rehashing everything.

Note: you could pool the ImmutableDictBuilder if you want to save that allocation.

Reply:
I'll just switch to an ImmutableDictBuilder here and see what the test insertion says.

Reply:
Insertions says the allocation of the dictionary builder vs using a pooled dictionary doesn't show up, so we'll just go with this approach